// @sapphire/plugin-subcommands/dist/lib/types/Events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubcommandPluginIdentifiers = exports.SubcommandPluginEvents = void 0;
exports.SubcommandPluginEvents = {
    ChatInputSubcommandRun: 'chatInputSubcommandRun',
    ChatInputSubcommandSuccess: 'chatInputSubcommandSuccess',
    ChatInputSubcommandNotFound: 'chatInputSubcommandNotFound',
    ChatInputSubcommandError: 'chatInputSubcommandError',
    MessageSubcommandRun: 'messageSubcommandRun',
    MessageSubcommandSuccess: 'messageSubcommandSuccess',
    MessageSubcommandNotFound: 'messageSubcommandNotFound',
    MessageSubcommandError: 'messageSubcommandError',
    SubcommandMappingIsMissingMessageCommandHandler: 'subcommandMappingIsMissingMessageCommandHandler',
    SubcommandMappingIsMissingChatInputCommandHandler: 'subcommandMappingIsMissingChatInputCommandHandler'
};
var SubcommandPluginIdentifiers;
(function (SubcommandPluginIdentifiers) {
    SubcommandPluginIdentifiers["MessageSubcommandNoMatch"] = "messageSubcommandNoMatch";
    SubcommandPluginIdentifiers["ChatInputSubcommandNoMatch"] = "chatInputSubcommandNoMatch";
    SubcommandPluginIdentifiers["SubcommandNotFound"] = "subcommandNotFound";
})(SubcommandPluginIdentifiers = exports.SubcommandPluginIdentifiers || (exports.SubcommandPluginIdentifiers = {}));
